File: Distribution-Grid-EV-CA/R_Yanning/05_02_sample_charging events_distance bins_IOU blocks.R
library(data.table)

home.event.block <- readRDS('data/mobility data/CSTDM processed/charge_event_new_block_home_sample42_draw12_split6699.rds')
wp.event.block <- readRDS('data/mobility data/CSTDM processed/charge_event_new_block_workpublic_sample42_draw12_split6699.rds')

block.to.feeder <- fread('data/mapping/block to feeder 2010/shape mapped/block_2010_to_feeder_map_line_combo.csv')

event.pool <- readRDS("data/charging data/charging_session_all_clean.rds")

IOU.blocks <- fread('data/mapping/block to feeder 2010/shape block IOUs++/shape block IOUs++.csv')

head(home.event.block)
head(wp.event.block)
head(block.to.feeder)

event.block.all <- rbind(home.event.block,wp.event.block)
event.block.all <- event.block.all[GEOID10%in%IOU.blocks$GEOID10]

event.block.all <- merge(event.block.all,block.to.feeder,by="GEOID10")

# add bins
event.block.all[,demand:=dist/3]
event.block.all[demand<=5,bin:=1]
event.block.all[demand>5 & demand<=10,bin:=2]
event.block.all[demand>10 & demand<=15,bin:=3]
event.block.all[demand>15 & demand<=20,bin:=4]
event.block.all[demand>20 & demand<=30,bin:=5]
event.block.all[demand>30 & demand<=50,bin:=6]
event.block.all[demand>50 & demand<=80,bin:=7]
event.block.all[demand>80,bin:=8]

event.pool[energy<=5,bin:=1]
event.pool[energy>5 & energy<=10,bin:=2]
event.pool[energy>10 & energy<=15,bin:=3]
event.pool[energy>15 & energy<=20,bin:=4]
event.pool[energy>20 & energy<=30,bin:=5]
event.pool[energy>30 & energy<=50,bin:=6]
event.pool[energy>50 & energy<=80,bin:=7]
event.pool[energy>80,bin:=8]

head(event.block.all)
head(event.pool)

############## sample with weight:level in public & housing in home ##############

### public: DC/L2 #######
level.weight <- data.table(level=c('DC','L2'),
                           ncharger=c(8919,29229))
n.event <- nrow(event.block.all[type=='P'])
set.seed(36)
event.block.all[,level:='none']
event.block.all[type=='P',
                level:=sample(level.weight$level,
                              n.event,
                              prob=level.weight$ncharger,
                              replace=T)]

nrow(event.block.all[level=='DC']) #3750433
nrow(event.block.all[level=='L2']) #12289247
# ratio very close to charger ratio

event.block.all[trip=='ld_corridor',level:='DC']
nrow(event.block.all[level=='DC']) #4125536
nrow(event.block.all[level=='L2']) #11914144
# ratio from 3.3 -> 2.9 (more DC added to corridor)

### home: single/multi #######
unique(event.pool$housing)
housing.weight <- data.table(house=c('single family','multi family'),
                             unit=c(56.4,38.9))
n.event <- nrow(event.block.all[type=='H'])

set.seed(36)
event.block.all[,house:='none']
event.block.all[type=='H',
                house:=sample(housing.weight$house,
                              n.event,
                              prob=housing.weight$unit,
                              replace=T)]

nrow(event.block.all[house=='single family']) #28712855
nrow(event.block.all[house=='multi family']) #19809304
# ratio very close to charger ratio

############## sample #########################################################
event.block.all[,n_event:=.N,keyby=.(type,level,house,bin)]
head(event.block.all)

event.pool[,eventID:=.I]
head(event.pool)

### home ###
start_time <- Sys.time()
for (h in c('single family','multi family')){
  for(b in 1:8){
    set.seed(36)
    event.block.all[type=='H' & house==h & bin==b,
                    eventID:=sample(event.pool[charge_type=='home' 
                                               & housing==h 
                                               & bin==b]$eventID,
                                    min(n_event),
                                    replace=T)]
  }
}
end_time <- Sys.time()
end_time - start_time # 4 secs

# summary(event.block.all$eventID)
# summary(event.block.all[type=='H']$eventID)

### public ###
start_time <- Sys.time()
for (l in c('DC','L2')){
  for(b in 1:8){
    set.seed(36)
    event.block.all[type=='P' & level==l & bin==b,
                    eventID:=sample(event.pool[charge_type=='public' 
                                               & charger_level==l 
                                               & bin==b]$eventID,
                                    min(n_event),
                                    replace=T)]
  }
}
end_time <- Sys.time()
end_time - start_time # 3 secs

### work ###
start_time <- Sys.time()
for(b in 1:8){
  set.seed(36)
  event.block.all[type=='W' & bin==b,
                  eventID:=sample(event.pool[charge_type=='work' 
                                             & bin==b]$eventID,
                                  min(n_event),
                                  replace=T)]
}
end_time <- Sys.time()
end_time - start_time # 2 secs

#summary(event.block.all$eventID) # all sampled

############## join empirical event into sampled IDs ####################################
event.feeder.sample <- event.pool[event.block.all,on=.(eventID)]
head(event.feeder.sample)

event.feeder.sample <- event.feeder.sample[,c('FeederID','year',
                                              'start_hour','end_hour',
                                              'charge_type','charger_level',
                                              'power','energy','housing')]

saveRDS(event.feeder.sample,'data/result/charging session by feeder/session_feeder_new_bins_IOU_sample42_draw12_split6699_sample36.rds')

File: Distribution-Grid-EV-CA/05_02_sample_charging_events_distance_bins_IOU_blocks.py
"""
05_02_sample_charging_events_distance_bins_IOU_blocks.py — Python port of
``05_02_sample_charging events_distance bins_IOU blocks.R``.

Samples empirical charging sessions (from ``charging_session_all_clean``) that
match each block-level event's distance bin / housing / level.
"""

from __future__ import annotations

import os

import numpy as np
import pandas as pd

from common import read_rds_like, save_rds_like

EVENT_HOME = (
    "data/mobility_data/CSTDM_processed/charge_event_new_block_home_sample42_draw12_split6699.pkl"
)
EVENT_WP = (
    "data/mobility_data/CSTDM_processed/charge_event_new_block_workpublic_sample42_draw12_split6699.pkl"
)
BLOCK_TO_FEEDER = "data/mapping/block to feeder 2010/shape mapped/block_2010_to_feeder_map_line_combo.csv"
IOU_BLOCKS = "data/mapping/block to feeder 2010/shape block IOUs++/shape block IOUs++.csv"
EVENT_POOL = "data/charging data/charging_session_all_clean.pkl"

OUT_PATH = (
    "data/result/charging session by feeder/"
    "session_feeder_new_bins_IOU_sample42_draw12_split6699_sample36.pkl"
)

DEMAND_BIN_EDGES = np.array([0, 5, 10, 15, 20, 30, 50, 80, np.inf])


def assign_bins(vec: pd.Series) -> pd.Series:
    """Returns bin IDs 1..8 following the R script's cutpoints."""
    return pd.cut(
        vec,
        bins=DEMAND_BIN_EDGES,
        labels=[1, 2, 3, 4, 5, 6, 7, 8],
        right=True,
        include_lowest=True,
    ).astype("Int64")


def main() -> None:
    home = read_rds_like(EVENT_HOME)
    wp = read_rds_like(EVENT_WP)
    block_to_feeder = pd.read_csv(BLOCK_TO_FEEDER)
    event_pool = read_rds_like(EVENT_POOL)
    iou_blocks = pd.read_csv(IOU_BLOCKS)

    event_block = pd.concat([home, wp], ignore_index=True)
    event_block = event_block[event_block["GEOID10"].isin(iou_blocks["GEOID10"])]
    event_block = pd.merge(event_block, block_to_feeder, on="GEOID10")

    event_block["demand"] = event_block["dist"] / 3.0
    event_block["bin"] = assign_bins(event_block["demand"])
    event_pool["bin"] = assign_bins(event_pool["energy"])

    # --- weighted sampling for public level / housing -----------------------
    level_weight = pd.DataFrame({"level": ["DC", "L2"], "ncharger": [8919, 29229]})
    rng = np.random.default_rng(36)

    event_block["level"] = "none"
    public_mask = event_block["type"] == "P"
    p_level = level_weight["ncharger"].to_numpy(dtype=float) / level_weight["ncharger"].sum()
    event_block.loc[public_mask, "level"] = rng.choice(
        level_weight["level"].to_numpy(),
        size=int(public_mask.sum()),
        replace=True,
        p=p_level,
    )
    event_block.loc[event_block["trip"] == "ld_corridor", "level"] = "DC"

    housing_weight = pd.DataFrame({"house": ["single family", "multi family"], "unit": [56.4, 38.9]})
    rng2 = np.random.default_rng(36)
    event_block["house"] = "none"
    home_mask = event_block["type"] == "H"
    p_h = housing_weight["unit"].to_numpy(dtype=float) / housing_weight["unit"].sum()
    event_block.loc[home_mask, "house"] = rng2.choice(
        housing_weight["house"].to_numpy(),
        size=int(home_mask.sum()),
        replace=True,
        p=p_h,
    )

    # --- sample event IDs from the empirical pool ----------------------------
    event_pool = event_pool.reset_index(drop=True).copy()
    event_pool["eventID"] = np.arange(1, len(event_pool) + 1)
    event_block["n_event"] = event_block.groupby(
        ["type", "level", "house", "bin"], dropna=False
    )["type"].transform("size")
    event_block["eventID"] = np.nan

    def _sample_ids(pool_mask: pd.Series, block_mask: pd.Series, seed: int = 36) -> None:
        sub_pool = event_pool.loc[pool_mask, "eventID"].to_numpy()
        n = int(block_mask.sum())
        if n == 0 or sub_pool.size == 0:
            return
        rng_local = np.random.default_rng(seed)
        ids = rng_local.choice(sub_pool, size=n, replace=True)
        event_block.loc[block_mask, "eventID"] = ids

    # home
    for h in ("single family", "multi family"):
        for b in range(1, 9):
            pool_mask = (
                (event_pool["charge_type"] == "home")
                & (event_pool["housing"] == h)
                & (event_pool["bin"] == b)
            )
            block_mask = (
                (event_block["type"] == "H") & (event_block["house"] == h) & (event_block["bin"] == b)
            )
            _sample_ids(pool_mask, block_mask)

    # public
    for lvl in ("DC", "L2"):
        for b in range(1, 9):
            pool_mask = (
                (event_pool["charge_type"] == "public")
                & (event_pool["charger_level"] == lvl)
                & (event_pool["bin"] == b)
            )
            block_mask = (
                (event_block["type"] == "P") & (event_block["level"] == lvl) & (event_block["bin"] == b)
            )
            _sample_ids(pool_mask, block_mask)

    # work
    for b in range(1, 9):
        pool_mask = (event_pool["charge_type"] == "work") & (event_pool["bin"] == b)
        block_mask = (event_block["type"] == "W") & (event_block["bin"] == b)
        _sample_ids(pool_mask, block_mask)

    # --- join empirical rows into each sampled event -------------------------
    event_feeder = pd.merge(event_block, event_pool, on="eventID", how="left")
    cols = [
        "FeederID", "year", "start_hour", "end_hour", "charge_type", "charger_level",
        "power", "energy", "housing",
    ]
    event_feeder = event_feeder[cols]
    save_rds_like(event_feeder, OUT_PATH)


if __name__ == "__main__":
    main()
